### ISSUE TITLE
When the perception is wrong, the logic still holds

### ISSUE THREAD

**Issue by JAYCHOU2020** [AbductiveLearning/ABL-HED#7]
I want to ask that when 5 pictures 1, +, 0, =, 1 are perceived as numbers and symbols 0, +, 1, =, 1, the logic layer gets that this is also an equation, which is still logically true. And then what? How does it optimize to get the pseudo-label back propagated to the perception model and finally get the right perception?

**Comment by LiuBodan** [comment]
@JAYCHOU2020 
Hi, I think the ABL model might accept the wrong perception for this iteration, as it is consistent with the domain knowledge, however, as training goes on and if we have a diverse training dataset, this wrong perception might cause more wrong perceptions which are inconsistent with the domain knowledge, and the logcial abduction will try to correct the incosistent ones and fixing the perception problem overtime.

I'm learning the ABL as well, you can add my discord rubber#4698 to have further discussion!

**Comment by Tony-HYX** [comment]
@JAYCHOU2020 
Hi,

Firstly, I would like to extend my sincere apologies for the late response. 

Regarding your query, I appreciate your insightful question. The logical consistency of the abduced label (0 + 1 = 1) in this context is treated as ground-truth label during the training process. However, it’s important to note that neural networks typically rely on batch training, where a batch of data is processed at once rather than individual data points. In the context of your example, it often happens that the majority of the labels in a batch are corrected and optimized for perception, this leads to an overall improvement in the performance of the neural network, as the correct labels will outweigh the incorrect ones.

The answer of @LiuBodan can also address this problem.

I hope this answers your query. If you have any further questions, please do not hesitate to reach out.

**Comment by AbductiveLearning** [comment]
🌟 **New!** [ABLkit](https://github.com/AbductiveLearning/ABLkit) released: A toolkit for Abductive Learning with high flexibility, user-friendly interface, and optimized performance. Welcome to try it out!🚀

Update: This repository is NO longer actively developed. It has been (mostly) superseded by [ABLkit](https://github.com/AbductiveLearning/ABLkit). The code for ABL-HED with ABLkit is in this [link](https://github.com/AbductiveLearning/ABLkit/tree/main/examples/hed). For the latest advancements and updates, we encourage you to visit the new repository.
